### ISSUE TITLE
no-else-return doesn't trigger for else if.

### ISSUE THREAD

**Issue by graingert** [eslint/eslint#9228]
**Tell us about your environment**

* **ESLint Version:** 4.6.1
* **Node Version:** 8.4.0
* **npm Version:** 5.3.0

**What parser (default, Babel-ESLint, etc.) are you using?** see config
**Please show your full configuration:** see eslint demo URLs.
**What did you do? Please include the actual source code causing the issue.**

```js
const res = (() => {
  if (error) {
    return 'It failed';
  } else if (loading) {
    return "It's still loading";
  }

  return result;
})();
```

**What did you expect to happen?** 

no-else-return should trigger.

**What actually happened? Please include the actual, raw output from ESLint.**

no-else-return did not trigger.

Demo urls:

[doesn't detect error](https://eslint.org/demo/#eyJ0ZXh0IjoiY29uc3QgcmVzID0gKCgpID0+IHtcbiAgaWYgKGVycm9yKSB7XG4gICAgcmV0dXJuICdJdCBmYWlsZWQnO1xuICB9IGVsc2UgaWYgKGxvYWRpbmcpIHtcbiAgICByZXR1cm4gXCJJdCdzIHN0aWxsIGxvYWRpbmdcIjtcbiAgfVxuICBcbiAgcmV0dXJuIHJlc3VsdDtcbn0pKCk7Iiwib3B0aW9ucyI6eyJwYXJzZXJPcHRpb25zIjp7ImVjbWFWZXJzaW9uIjo4LCJzb3VyY2VUeXBlIjoic2NyaXB0IiwiZWNtYUZlYXR1cmVzIjp7fX0sInJ1bGVzIjp7Im5vLWVsc2UtcmV0dXJuIjoyfSwiZW52Ijp7fX19)
[does detect error](https://eslint.org/demo/#eyJ0ZXh0IjoiY29uc3QgcmVzID0gKCgpID0+IHtcbiAgaWYgKGVycm9yKSB7XG4gICAgcmV0dXJuICdJdCBmYWlsZWQnO1xuICB9IGVsc2Uge1xuICAgIGlmIChsb2FkaW5nKSB7XG4gICAgICByZXR1cm4gXCJJdCdzIHN0aWxsIGxvYWRpbmdcIjtcbiAgICB9XG4gIH1cblxuICBcbiAgcmV0dXJuIHJlc3VsdDtcbn0pKCk7Iiwib3B0aW9ucyI6eyJwYXJzZXJPcHRpb25zIjp7ImVjbWFWZXJzaW9uIjo4LCJzb3VyY2VUeXBlIjoic2NyaXB0IiwiZWNtYUZlYXR1cmVzIjp7fX0sInJ1bGVzIjp7Im5vLWVsc2UtcmV0dXJuIjoyfSwiZW52Ijp7fX19)

**Comment by platinumazure** [comment]
I believe this is working as designed. With an `else if`, there is another code path (both conditionals not met), which results in an implicit `return undefined`. Additionally, the rule is designed to call out `else` blocks that could be removed (around return statements) without changing the meaning of the code. Removing `else if` changes the meaning of the code.

**Comment by graingert** [comment]
@platinumazure not in the example given.

**Comment by graingert** [comment]
@platinumazure not in the example I gave.

**Comment by graingert** [comment]
@platinumazure can you give a code example of what you mean and I'll add it to my tests?

**Comment by platinumazure** [comment]
No, I see the hook in your example. I'm not sure how easy it will be to enhance the rule for your case without potentially breaking other cases, but that's why we have unit tests.

**Comment by graingert** [comment]
The hook?

**Comment by graingert** [comment]
> I believe this is working as designed. With an else if, there is another code path (both conditionals not met), which results in an implicit return undefined. Additionally, the rule is designed to call out else blocks that could be removed (around return statements) without changing the meaning of the code. Removing else if changes the meaning of the code.

I'm pretty sure that's not true. If you transform the code to explicit block statements no-else-return [detects it correctly](https://eslint.org/demo/#eyJ0ZXh0IjoiY29uc3QgcmVzID0gKCgpID0+IHtcbiAgaWYgKGVycm9yKSB7XG4gICAgcmV0dXJuICdJdCBmYWlsZWQnO1xuICB9IGVsc2Uge1xuICAgIGlmIChsb2FkaW5nKSB7XG4gICAgICByZXR1cm4gXCJJdCdzIHN0aWxsIGxvYWRpbmdcIjtcbiAgICB9XG4gIH1cblxuICBcbiAgcmV0dXJuIHJlc3VsdDtcbn0pKCk7Iiwib3B0aW9ucyI6eyJwYXJzZXJPcHRpb25zIjp7ImVjbWFWZXJzaW9uIjo4LCJzb3VyY2VUeXBlIjoic2NyaXB0IiwiZWNtYUZlYXR1cmVzIjp7fX0sInJ1bGVzIjp7Im5vLWVsc2UtcmV0dXJuIjoyfSwiZW52Ijp7fX19)

**Comment by platinumazure** [comment]
Sorry- I see what I had missed. I see why you are asking if the rule should be enhanced.

I'm worried about what rabbit hole we might be opening up if we expand these cases too much.

I guess in general, we could conceivably flag all `else if` statements with a single return statement, but maybe only working our way from the bottom up?

```js
if (foo) return true;
else if (bar) { return true; }   // can't be flagged due to next line?
else if (baz) { doSomething(); return true; }
else if (x) return true;  // could be flagged?
else if (y) { return true; } // could be flagged?
```

I think the team might just decide that the rule is only designed to handle the specific case of `else` followed by `return`.

**Comment by platinumazure** [comment]
> I'm pretty sure that's not true. If you transform the code to explicit block statements no-else-return detects it correctly

So it does. I stand corrected.

I think we can accept this as a bug. Please feel free to continue work on your PR and thanks for your patience!

**Comment by graingert** [comment]
@platinumazure fyi, https://eslint.org/demo/#eyJ0ZXh0IjoiKCkgPT4ge1xuICBpZiAoZm9vKSByZXR1cm4gdHJ1ZTtcbiAgZWxzZSB7XG4gICAgaWYgKGJhcikgeyByZXR1cm4gdHJ1ZTsgfSAgIC8vIGNhbid0IGJlIGZsYWdnZWQgZHVlIHRvIG5leHQgbGluZT9cbiAgICBlbHNlIHtcbiAgICAgIGlmIChiYXopIHsgZG9Tb21ldGhpbmcoKTsgcmV0dXJuIHRydWU7IH1cbiAgICAgICAgZWxzZSB7XG4gICAgICAgICAgaWYgKHgpIHJldHVybiB0cnVlOyAgLy8gY291bGQgYmUgZmxhZ2dlZD9cbiAgICAgICAgICAgIGVsc2Uge1xuICAgICAgICAgICAgICBpZiAoeSkgeyByZXR1cm4gdHJ1ZTsgfSAvLyBjb3VsZCBiZSBmbGFnZ2VkP1xuICAgICAgICAgICAgfVxuICAgICAgICB9XG4gICAgfVxuICB9XG59OyIsIm9wdGlvbnMiOnsicGFyc2VyT3B0aW9ucyI6eyJlY21hVmVyc2lvbiI6OCwic291cmNlVHlwZSI6InNjcmlwdCIsImVjbWFGZWF0dXJlcyI6e319LCJydWxlcyI6eyJuby1lbHNlLXJldHVybiI6Mn0sImVudiI6e319fQ== flags them all.

**Comment by graingert** [comment]
```js
() => {
  if (foo) {
    return true;
  }

  if (bar) {
    return true;
  }

  if (baz) {
    doSomething();
    return true;
  }

  if (x) {
    return true;
  }

  if (y) {
    return true;
  }
}
```

**Comment by graingert** [comment]
@platinumazure is there a way to run only one test? Currently it takes too long to drop to a debugger on one rule test.

**Comment by platinumazure** [comment]
@graingert Use Mocha's `--grep` flag to dig down to a specific test (and make sure you run Mocha against just the one test file, if you aren't already).

**Comment by graingert** [comment]
@platinumazure how do I get it to skip running eslint?

**Comment by platinumazure** [comment]
Just run Mocha directly:

```bash
./node-modules/mocha/bin/_mocha tests/lib/rules/no-else-return.js --grep "Name of test"
```

**Comment by graingert** [comment]
@platinumazure thanks, that was super handy.

**Comment by graingert** [comment]
@platinumazure I've taken my PR out of WIP and it's ready for review.
